File: codes/julia/ch14/ho_pseudospectra.jl
# ho_pseudospectra.jl
#
# Chapter 14: Higher-Order Boundary Value Problems
#
# Computational Etude 14.8: Pseudospectra of the Orr-Sommerfeld Operator.
#
# Computes the epsilon-pseudospectra of the Orr-Sommerfeld operator for
# plane Poiseuille flow at the critical Reynolds number R = 5772,
# streamwise wavenumber alpha = 1, using N = 100 Chebyshev points.
#
# The epsilon-pseudospectrum of the matrix pencil (A, B) is:
#
#     Lambda_eps = { z in C : sigma_min(A - z B) < eps }
#
# For normal operators the pseudospectra are simply eps-balls around the
# eigenvalues. The Orr-Sommerfeld operator is highly non-normal, so
# its pseudospectra extend far from the eigenvalues, revealing extreme
# sensitivity to perturbations.
#
# At R = 5772 (the critical Reynolds number for plane Poiseuille flow),
# all eigenvalues lie barely in the stable half-plane. The contour plot
# of log10(sigma_min) shows that eps-perturbations of size as small as
# 10^{-8} can push eigenvalues into the unstable half-plane, explaining
# the subcritical transition to turbulence.
#
# Reference: Trefethen, "Spectral Methods in MATLAB", Program 40.
#            Trefethen & Embree, "Spectra and Pseudospectra" (2005).
#            Reddy, Schmid & Henningson, SIAM J. Appl. Math. 53 (1993).
#
# Generates Figure 14.7: Pseudospectra contour plot with eigenvalues.
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LinearAlgebra
using Printf

# Import Chebyshev matrix function from Chapter 7
include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

# Publication-quality CairoMakie configuration
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# Book color scheme
const NAVY   = colorant"#142D6E"
const SKY    = colorant"#7896D2"
const CORAL  = colorant"#E74C3C"
const TEAL   = colorant"#16A085"
const PURPLE = colorant"#8E44AD"
const ORANGE = colorant"#E67E22"

# Output paths
const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch14", "julia")

# -----------------------------------------------------------------------------
# Orr-Sommerfeld matrices (boundary-bordering technique)
# -----------------------------------------------------------------------------
"""
    build_orr_sommerfeld_matrices(N, R, alpha)

Build the generalized eigenvalue matrices A, B for the Orr-Sommerfeld
equation for plane Poiseuille flow U(y) = 1 - y^2.

Boundary conditions: v(+-1) = v'(+-1) = 0 (no-slip).
Dirichlet conditions v(+-1) = 0 are enforced by restriction to interior
points (indices 2:N). Neumann conditions v'(+-1) = 0 are imposed by
replacing the first and last rows of both A and B.

Returns A, B matrices of size (N-1) x (N-1).
"""
function build_orr_sommerfeld_matrices(N::Int, R::Float64, alpha::Float64)
    D, x = cheb_matrix(N)

    # Second and fourth derivative matrices
    D2_full = D * D
    D4_full = D2_full * D2_full

    # Interior grid (remove boundary points x[1] = 1, x[N+1] = -1)
    x_int = x[2:N]

    # Base flow and its second derivative
    # U(y) = 1 - y^2,  U''(y) = -2
    U = Diagonal(1.0 .- x_int.^2)
    U2 = -2.0 * I(N - 1)    # diag(U'')

    # Interior blocks of D^2 and D^4
    D2_int = D2_full[2:N, 2:N]
    D4_int = D4_full[2:N, 2:N]

    # Shift operator: S = D^2 - alpha^2 I
    M_dim = N - 1
    I_int = Matrix(1.0I, M_dim, M_dim)
    S = D2_int - alpha^2 * I_int

    # S^2 = D^4 - 2 alpha^2 D^2 + alpha^4 I
    S2 = D4_int - 2.0 * alpha^2 * D2_int + alpha^4 * I_int

    # A = S^2 / (i alpha R) - U S + U''
    # B = -S
    A = S2 / (1im * alpha * R) - U * S + U2
    B = -copy(S)

    # Impose Neumann boundary conditions via row replacement:
    # v'(+1) = D[1, 2:N] @ v_int = 0  => row 1
    # v'(-1) = D[N+1, 2:N] @ v_int = 0  => row M_dim
    A[1, :] = D[1, 2:N]
    B[1, :] .= 0.0

    A[M_dim, :] = D[N+1, 2:N]
    B[M_dim, :] .= 0.0

    return A, B
end


# -----------------------------------------------------------------------------
# Eigenvalue computation
# -----------------------------------------------------------------------------
"""
    compute_eigenvalues(A, B)

Solve A v = lambda B v, filter spurious eigenvalues, sort by Im(lambda).
"""
function compute_eigenvalues(A, B)
    eigenvalues = eigvals(A, B)

    # Filter out infinite and NaN eigenvalues
    finite_mask = isfinite.(eigenvalues) .& (abs.(eigenvalues) .< 1e8)
    eigenvalues = eigenvalues[finite_mask]

    # Sort by imaginary part (most unstable first)
    idx = sortperm(imag.(eigenvalues), rev=true)
    eigenvalues = eigenvalues[idx]

    return eigenvalues
end


# -----------------------------------------------------------------------------
# Pseudospectra computation
# -----------------------------------------------------------------------------
"""
    compute_pseudospectra(A, B, x_re, x_im)

Compute log10(sigma_min(A - z B)) on a grid of complex numbers z.

For the generalized eigenvalue problem A v = lambda B v, the
epsilon-pseudospectrum is the set of z where sigma_min(A - z B) < eps.
"""
function compute_pseudospectra(A, B, x_re, x_im)
    nx = length(x_re)
    ny = length(x_im)

    sigma_grid = zeros(ny, nx)

    total = nx * ny
    count = 0

    for j in 1:ny
        for i in 1:nx
            z = x_re[i] + 1im * x_im[j]
            # Compute smallest singular value of (A - z B)
            sv = svdvals(A - z * B)
            sigma_grid[j, i] = log10(max(sv[end], 1e-20))
            count += 1
        end

        # Progress report every 10 rows
        if j % 10 == 0
            pct = 100.0 * count / total
            @printf("    ... %.0f%% complete (%d/%d grid points)\n", pct, count, total)
        end
    end

    return sigma_grid
end


# -----------------------------------------------------------------------------
# Main
# -----------------------------------------------------------------------------
function main()
    mkpath(OUTPUT_DIR)

    # Physical parameters
    R = 5772.0         # Critical Reynolds number for plane Poiseuille flow
    alpha = 1.0        # Streamwise wavenumber
    N = 100            # Number of Chebyshev intervals

    println("Pseudospectra of the Orr-Sommerfeld Operator")
    println("=" ^ 60)
    @printf("  Reynolds number:      R = %d\n", Int(R))
    @printf("  Wavenumber:           alpha = %.0f\n", alpha)
    @printf("  Chebyshev points:     N = %d\n", N)
    @printf("  Matrix size:          %d x %d\n", N - 1, N - 1)
    println()

    # ---- Build matrices and solve eigenvalue problem ----
    println("  Building Orr-Sommerfeld matrices ...")
    A, B = build_orr_sommerfeld_matrices(N, R, alpha)
    @printf("  Matrix shapes: A = %s, B = %s\n", string(size(A)), string(size(B)))

    println("  Computing eigenvalues ...")
    eigenvalues = compute_eigenvalues(A, B)
    @printf("  Number of finite eigenvalues: %d\n", length(eigenvalues))

    if length(eigenvalues) > 0
        # The most unstable eigenvalue has the largest Im(lambda)
        lam_crit = eigenvalues[1]
        @printf("  Most unstable eigenvalue:\n")
        @printf("    lambda = %.10f + %.10fi\n", real(lam_crit), imag(lam_crit))
        if imag(lam_crit) < 0
            println("    (stable: Im(lambda) < 0)")
        else
            println("    (UNSTABLE: Im(lambda) > 0)")
        end
    end

    # Print a few leading eigenvalues
    n_show = min(10, length(eigenvalues))
    @printf("\n  Leading %d eigenvalues (sorted by Im):\n", n_show)
    @printf("  %16s  %16s\n", "Re(lambda)", "Im(lambda)")
    println("  " * "-" ^ 36)
    for k in 1:n_show
        @printf("  %16.10f  %16.10f\n", real(eigenvalues[k]), imag(eigenvalues[k]))
    end
    println()

    # ---- Define grid for pseudospectra ----
    nx, ny = 120, 120
    x_re = range(-0.8, 0.2, length=nx)
    x_im = range(-1.0, 0.05, length=ny)

    @printf("  Computing pseudospectra on %d x %d grid ...\n", nx, ny)
    @printf("  Grid: Re in [%.1f, %.1f], Im in [%.2f, %.2f]\n",
            x_re[1], x_re[end], x_im[1], x_im[end])
    @printf("  This requires %d SVD computations of %dx%d matrices.\n",
            nx * ny, N - 1, N - 1)
    println()

    t_start = time()
    sigma_grid = compute_pseudospectra(A, B, x_re, x_im)
    t_elapsed = time() - t_start
    @printf("\n  Pseudospectra computed in %.1f seconds.\n\n", t_elapsed)

    # ---- Plot ----
    println("  Generating figure ...")

    fig = Figure(size = (700, 560))

    ax = Axis(fig[1, 1],
              xlabel = L"\mathrm{Re}(\lambda)",
              ylabel = L"\mathrm{Im}(\lambda)",
              title = @sprintf("\\varepsilon-Pseudospectra: R = %d, \\alpha = %g, N = %d",
                               Int(R), alpha, N))

    # Contour levels: log10(eps) = -1, -2, ..., -8
    levels = collect(-8.0:1.0:-1.0)

    # Filled contours
    cf = contourf!(ax, collect(x_re), collect(x_im), sigma_grid,
                   levels = levels, colormap = :YlOrRd, extendlow = :auto,
                   extendhigh = :auto)

    # Contour lines for clarity
    contour!(ax, collect(x_re), collect(x_im), sigma_grid,
             levels = levels, color = :black, linewidth = 0.5, alpha = 0.6)

    Colorbar(fig[1, 2], cf,
             label = L"\log_{10}\,\sigma_{\min}(C - z\,I)",
             ticks = levels)

    # Overlay eigenvalues (only those within the plot region)
    in_region = (real.(eigenvalues) .>= x_re[1]) .& (real.(eigenvalues) .<= x_re[end]) .&
                (imag.(eigenvalues) .>= x_im[1]) .& (imag.(eigenvalues) .<= x_im[end])
    eigs_plot = eigenvalues[in_region]

    scatter!(ax, real.(eigs_plot), imag.(eigs_plot),
             color = :white, markersize = 5,
             strokecolor = NAVY, strokewidth = 0.8)

    # Stability boundary: Im(lambda) = 0
    hlines!(ax, [0.0], color = :gray, linestyle = :dash, linewidth = 0.8, alpha = 0.5)

    xlims!(ax, x_re[1], x_re[end])
    ylims!(ax, x_im[1], x_im[end])

    # Save
    save(joinpath(OUTPUT_DIR, "pseudospectra.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "pseudospectra.png"), fig, px_per_unit = 4)
    println("  Figure saved to: $(joinpath(OUTPUT_DIR, "pseudospectra.pdf"))")

    println()
    println("=" ^ 60)
end

main()

File: codes/python/ch14/ho_pseudospectra.py
#!/usr/bin/env python3
"""
ho_pseudospectra.py
Chapter 14: Higher-Order Boundary Value Problems

Computational Etude 14.8: Pseudospectra of the Orr-Sommerfeld Operator.

Computes the epsilon-pseudospectra of the Orr-Sommerfeld operator for
plane Poiseuille flow at the critical Reynolds number R = 5772,
streamwise wavenumber alpha = 1, using N = 100 Chebyshev points.

The Orr-Sommerfeld equation governs the linear stability of parallel
shear flows.  For the base flow U(y) = 1 - y^2 on [-1, 1], the
generalized eigenvalue problem reads A v = lambda B v, where A and B
encode the fourth-order ODE with no-slip boundary conditions
v(+-1) = v'(+-1) = 0.

The Dirichlet conditions are handled by restriction to interior
Chebyshev points.  The derivative (Neumann) conditions v'(+-1) = 0
are imposed by boundary bordering.  The resulting two algebraic
constraints are then eliminated analytically via a projection that
expresses the boundary unknowns in terms of the interior unknowns,
yielding a non-singular reduced system whose matrix C = B_red^{-1} A_red
is the one whose pseudospectra we compute.

The epsilon-pseudospectrum is defined as

    Lambda_eps(C) = { z in C : sigma_min(C - z I) < eps }

where sigma_min denotes the smallest singular value.  For normal
operators the pseudospectra are simply eps-balls around the
eigenvalues.  The Orr-Sommerfeld operator is highly non-normal, so
its pseudospectra extend far from the eigenvalues, revealing extreme
sensitivity to perturbations.

At R = 5772 (the critical Reynolds number for plane Poiseuille flow),
all eigenvalues lie barely in the stable half-plane Im(lambda) < 0.
The contour plot of log10(sigma_min) shows that perturbations of size
as small as 10^{-8} can push eigenvalues across the stability boundary
into Im(lambda) > 0, explaining the subcritical transition to
turbulence observed in pipe-flow experiments.

Reference: Trefethen, "Spectral Methods in MATLAB", Program 40.
           Trefethen & Embree, "Spectra and Pseudospectra" (2005).
           Reddy, Schmid & Henningson, SIAM J. Appl. Math. 53 (1993).

Author: Dr. Denys Dutykh
        Mathematics Department
        Khalifa University of Science and Technology
        Abu Dhabi, UAE

Last modified: 2026-02-22

Part of the book "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""

import numpy as np
from scipy.linalg import svdvals
import matplotlib.pyplot as plt
from matplotlib import rcParams
from pathlib import Path
import time

# Import Chebyshev matrix function from Chapter 7
import sys
sys.path.insert(0, str(Path(__file__).parent.parent / 'ch07'))
from cheb_matrix import cheb_matrix

# -----------------------------------------------------------------------------
# Publication-quality matplotlib configuration
# -----------------------------------------------------------------------------
rcParams['font.family'] = 'serif'
rcParams['font.serif'] = ['Computer Modern Roman', 'CMU Serif', 'DejaVu Serif']
rcParams['mathtext.fontset'] = 'cm'
rcParams['axes.labelsize'] = 11
rcParams['axes.titlesize'] = 12
rcParams['xtick.labelsize'] = 10
rcParams['ytick.labelsize'] = 10
rcParams['legend.fontsize'] = 10
rcParams['figure.dpi'] = 150
rcParams['savefig.dpi'] = 300
rcParams['text.usetex'] = False
rcParams['axes.linewidth'] = 0.8
rcParams['xtick.major.width'] = 0.8
rcParams['ytick.major.width'] = 0.8

# -----------------------------------------------------------------------------
# Book color scheme
# -----------------------------------------------------------------------------
NAVY = '#142D6E'
SKY = '#7896D2'
CORAL = '#E74C3C'
TEAL = '#16A085'
PURPLE = '#8E44AD'
ORANGE = '#E67E22'

# Output paths
SCRIPT_DIR = Path(__file__).parent
OUTPUT_DIR = SCRIPT_DIR / '..' / '..' / '..' / 'textbook' / 'figures' / 'ch14' / 'python'


# -----------------------------------------------------------------------------
# Build Orr-Sommerfeld matrix C = B^{-1} A via constraint elimination
# -----------------------------------------------------------------------------
def build_orr_sommerfeld_matrix(N, R, alpha):
    """
    Build the matrix C = B^{-1} A for the Orr-Sommerfeld eigenvalue
    problem for plane Poiseuille flow U(y) = 1 - y^2.

    The Orr-Sommerfeld equation, written as a generalized eigenvalue
    problem A v = lambda B v, is:

        A = (D^2 - alpha^2)^2 / (i alpha R)
            - diag(U) (D^2 - alpha^2) + diag(U'')
        B = -(D^2 - alpha^2)

    restricted to interior Chebyshev points (Dirichlet BCs v(+-1) = 0).

    The Neumann conditions v'(+-1) = 0 are imposed by replacing the
    first and last rows with derivative constraints, making B singular.
    These two algebraic constraints are then eliminated analytically:
    the constraint rows express the boundary unknowns (v at interior
    points nearest +-1) as linear combinations of the remaining
    unknowns.  Substituting into the dynamical rows yields a reduced
    system with non-singular B_red, allowing C = B_red^{-1} A_red.

    Parameters
    ----------
    N : int
        Number of Chebyshev intervals (N+1 grid points, N-1 interior).
    R : float
        Reynolds number.
    alpha : float
        Streamwise wavenumber.

    Returns
    -------
    C : ndarray, shape (N-3, N-3)
        Matrix C = B_red^{-1} A_red for pseudospectral analysis.
    eigenvalues : ndarray
        Eigenvalues of C, sorted by imaginary part (descending).
    """
    D, x = cheb_matrix(N)

    # Second and fourth derivative matrices
    D2 = D @ D
    D4 = D2 @ D2

    # Number of interior points
    n = N - 1

    # Interior grid (remove boundary points x_0 = 1, x_N = -1)
    x_int = x[1:N]

    # Base flow: U(y) = 1 - y^2,  U''(y) = -2
    U = np.diag(1.0 - x_int**2)
    U2 = -2.0 * np.eye(n)

    # Interior blocks of D^2 and D^4
    D2_int = D2[1:N, 1:N]
    D4_int = D4[1:N, 1:N]

    # Shift operator: S = D^2 - alpha^2 I
    I_int = np.eye(n)
    S = D2_int - alpha**2 * I_int

    # S^2 = D^4 - 2 alpha^2 D^2 + alpha^4 I
    S2 = D4_int - 2.0 * alpha**2 * D2_int + alpha**4 * I_int

    # Build A and B (boundary-bordering form)
    A = S2 / (1j * alpha * R) - U @ S + U2
    B = -S.copy()

    # Replace first/last rows with Neumann conditions v'(+-1) = 0
    A[0, :] = D[0, 1:N]
    B[0, :] = 0.0
    A[-1, :] = D[N, 1:N]
    B[-1, :] = 0.0

    # --- Constraint elimination ---
    # Rows 0 and n-1 are algebraic constraints: A[0,:] v = 0, A[-1,:] v = 0
    # (since B[0,:] = B[-1,:] = 0, the eigenvalue lambda does not appear).
    #
    # Partition v = [v_0; v_free; v_{n-1}] where v_free = v[1:-1].
    # The constraints give:
    #   A[0,0]  v_0 + A[0,1:-1]  v_free + A[0,-1]  v_{n-1} = 0
    #   A[-1,0] v_0 + A[-1,1:-1] v_free + A[-1,-1] v_{n-1} = 0
    #
    # Solve for [v_0; v_{n-1}] = P @ v_free where P = -CC^{-1} F.

    CC = np.array([[A[0, 0], A[0, -1]],
                   [A[-1, 0], A[-1, -1]]])
    F = np.vstack([A[0, 1:-1], A[-1, 1:-1]])
    P = -np.linalg.solve(CC, F)      # shape (2, n-2)

    # Build the lifting matrix T: v = T @ v_free, shape (n, n-2)
    T = np.zeros((n, n - 2), dtype=complex)
    T[0, :] = P[0, :]                # v_0 row
    T[1:-1, :] = np.eye(n - 2)       # v_free rows
    T[-1, :] = P[1, :]               # v_{n-1} row

    # Reduced system: use the dynamical rows (1 to n-2) of A, B
    A_red = A[1:-1, :] @ T           # shape (n-2, n-2)
    B_red = B[1:-1, :] @ T           # shape (n-2, n-2)

    # Form C = B_red^{-1} A_red (B_red is now non-singular)
    C = np.linalg.solve(B_red, A_red)

    # Eigenvalues
    eigenvalues = np.linalg.eigvals(C)
    idx = np.argsort(-eigenvalues.imag)
    eigenvalues = eigenvalues[idx]

    return C, eigenvalues


# -----------------------------------------------------------------------------
# Pseudospectra computation
# -----------------------------------------------------------------------------
def compute_pseudospectra(C, x_re, x_im):
    """
    Compute log10(sigma_min(C - z I)) on a grid of complex numbers z.

    For each point z = x + iy on the grid, computes the smallest
    singular value of (C - zI).  The level set where this equals eps
    is the boundary of the eps-pseudospectrum.

    Parameters
    ----------
    C : ndarray, shape (M, M)
        The matrix whose pseudospectra are computed.
    x_re : ndarray, shape (nx,)
        Real parts of the grid points.
    x_im : ndarray, shape (ny,)
        Imaginary parts of the grid points.

    Returns
    -------
    sigma_grid : ndarray, shape (ny, nx)
        log10 of the minimum singular value at each grid point.
    """
    nx = len(x_re)
    ny = len(x_im)
    M = C.shape[0]
    I = np.eye(M, dtype=C.dtype)

    sigma_grid = np.zeros((ny, nx))

    total = nx * ny
    count = 0

    for j in range(ny):
        for i in range(nx):
            z = x_re[i] + 1j * x_im[j]
            # Compute smallest singular value of (C - z I)
            sv = svdvals(C - z * I)
            sigma_grid[j, i] = np.log10(max(sv[-1], 1e-20))
            count += 1

        # Progress report every 10 rows
        if (j + 1) % 10 == 0:
            pct = 100.0 * count / total
            print(f"    ... {pct:.0f}% complete ({count}/{total} grid points)")

    return sigma_grid


# -----------------------------------------------------------------------------
# Plotting
# -----------------------------------------------------------------------------
def plot_pseudospectra(x_re, x_im, sigma_grid, eigenvalues, R, alpha, N):
    """
    Plot the pseudospectra as filled contours with eigenvalues overlaid.

    Parameters
    ----------
    x_re, x_im : ndarray
        Grid coordinates.
    sigma_grid : ndarray
        log10(sigma_min) values on the grid.
    eigenvalues : ndarray
        Eigenvalues of C (complex).
    R : float
        Reynolds number.
    alpha : float
        Streamwise wavenumber.
    N : int
        Number of Chebyshev intervals.

    Returns
    -------
    fig : matplotlib Figure
    """
    fig, ax = plt.subplots(figsize=(8, 6.5))

    # Contour levels: log10(eps) = -1, -2, ..., -8
    levels = np.arange(-8, 0)

    # Filled contours
    cf = ax.contourf(x_re, x_im, sigma_grid, levels=levels, cmap='YlOrRd_r',
                     extend='both')

    # Contour lines (black) for clarity
    cl = ax.contour(x_re, x_im, sigma_grid, levels=levels, colors='k',
                    linewidths=0.5, alpha=0.6)
    ax.clabel(cl, inline=True, fontsize=7, fmt='%d')

    # Colorbar
    cbar = fig.colorbar(cf, ax=ax, shrink=0.85, pad=0.02)
    cbar.set_label(r'$\log_{10}\,\sigma_{\min}(C - z\,I)$', fontsize=11)
    cbar.set_ticks(levels)
    cbar.set_ticklabels([str(lev) for lev in levels])

    # Overlay eigenvalues (only those within the plot region)
    in_region = ((eigenvalues.real >= x_re[0]) & (eigenvalues.real <= x_re[-1]) &
                 (eigenvalues.imag >= x_im[0]) & (eigenvalues.imag <= x_im[-1]))
    eigs_plot = eigenvalues[in_region]

    ax.plot(eigs_plot.real, eigs_plot.imag, 'o', color='white',
            markersize=4, markeredgecolor=NAVY, markeredgewidth=0.8,
            zorder=10, label='Eigenvalues')

    # Stability boundary: Im(lambda) = 0
    ax.axhline(0, color='gray', linestyle='--', linewidth=0.8, alpha=0.5)

    ax.set_xlabel(r'$\mathrm{Re}(\lambda)$')
    ax.set_ylabel(r'$\mathrm{Im}(\lambda)$')
    ax.set_title(
        r'$\varepsilon$-Pseudospectra of Orr--Sommerfeld: '
        r'$R = %d$, $\alpha = %g$, $N = %d$' % (R, alpha, N),
        fontsize=12
    )
    ax.legend(loc='upper left', fontsize=9, framealpha=0.9)

    ax.set_xlim(x_re[0], x_re[-1])
    ax.set_ylim(x_im[0], x_im[-1])

    return fig


# -----------------------------------------------------------------------------
# Main
# -----------------------------------------------------------------------------
def main():
    """Compute and plot pseudospectra of the Orr-Sommerfeld operator."""

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    # Physical parameters
    R = 5772           # Critical Reynolds number for plane Poiseuille flow
    alpha = 1.0        # Streamwise wavenumber
    N = 100            # Number of Chebyshev intervals

    print("Pseudospectra of the Orr-Sommerfeld Operator")
    print("=" * 60)
    print(f"  Reynolds number:      R = {R}")
    print(f"  Wavenumber:           alpha = {alpha}")
    print(f"  Chebyshev points:     N = {N}")
    print(f"  Interior points:      {N - 1}")
    print(f"  Matrix C size:        {N - 3} x {N - 3} (after constraint elimination)")
    print()

    # ---- Build reduced matrix C ----
    print("  Building Orr-Sommerfeld matrix C = B_red^{-1} A_red ...")
    C, eigenvalues = build_orr_sommerfeld_matrix(N, R, alpha)
    print(f"  Matrix C shape: {C.shape}")
    print(f"  Number of eigenvalues: {len(eigenvalues)}")

    if len(eigenvalues) > 0:
        lam_crit = eigenvalues[0]
        print(f"  Most unstable eigenvalue:")
        print(f"    lambda = {lam_crit.real:.10f} + {lam_crit.imag:.10f} i")
        if lam_crit.imag < 0:
            print(f"    (stable: Im(lambda) < 0)")
        else:
            print(f"    (UNSTABLE: Im(lambda) > 0)")

    # Print leading eigenvalues
    n_show = min(10, len(eigenvalues))
    print(f"\n  Leading {n_show} eigenvalues (sorted by Im):")
    print(f"  {'Re(lambda)':>16s}  {'Im(lambda)':>16s}")
    print("  " + "-" * 36)
    for k in range(n_show):
        print(f"  {eigenvalues[k].real:>16.10f}  {eigenvalues[k].imag:>16.10f}")
    print()

    # ---- Define grid for pseudospectra ----
    # The eigenvalues lie in Re ~ [0.2, 1.0], Im ~ [-1.0, 0],
    # so we center the grid on this region with some padding.
    nx, ny = 120, 120
    x_re = np.linspace(-0.1, 1.1, nx)
    x_im = np.linspace(-1.0, 0.1, ny)

    print(f"  Computing pseudospectra on {nx} x {ny} grid ...")
    print(f"  Grid: Re in [{x_re[0]}, {x_re[-1]}], Im in [{x_im[0]}, {x_im[-1]}]")
    print(f"  This requires {nx * ny} SVD computations of {C.shape[0]}x{C.shape[0]} matrices.")
    print()

    t_start = time.time()
    sigma_grid = compute_pseudospectra(C, x_re, x_im)
    t_elapsed = time.time() - t_start
    print(f"\n  Pseudospectra computed in {t_elapsed:.1f} seconds.")
    print()

    # ---- Plot ----
    print("  Generating figure ...")
    fig = plot_pseudospectra(x_re, x_im, sigma_grid, eigenvalues, R, alpha, N)

    plt.tight_layout()

    # Save
    output_file = OUTPUT_DIR / 'pseudospectra.pdf'
    fig.savefig(output_file, bbox_inches='tight', pad_inches=0.05)
    fig.savefig(output_file.with_suffix('.png'), bbox_inches='tight',
                pad_inches=0.05, dpi=300)
    print(f"  Figure saved to: {output_file.resolve()}")
    plt.close(fig)

    print()
    print("=" * 60)


if __name__ == '__main__':
    main()
